Applications/Spire/Source/Ui/Box.cpp

```diff
@@ -112,174 +321,285 @@ bool Box::event(QEvent* event) {
+        },
+        [&] (std::in_place_type_t<PaddingTop>, int size) {
+          m_body_geometry.setTop(m_body_geometry.top() + size);
+        },
+        [&] (std::in_place_type_t<PaddingRight>, int size) {
+          m_body_geometry.setRight(m_body_geometry.right() - size);
+        },
+        [&] (std::in_place_type_t<PaddingBottom>, int size) {
+          m_body_geometry.setBottom(m_body_geometry.bottom() - size);
+        },
+        [&] (std::in_place_type_t<PaddingLeft>, int size) {
+          m_body_geometry.setLeft(m_body_geometry.left() + size);
+        });
+    }
+    m_container->setGeometry(m_body_geometry);
   }
+  update_border_geometry();
   QWidget::resizeEvent(event);
 }
 
-void Box::commit_style() {
-  auto stylesheet = QString(
-    R"(#0x%1 {
-        border-style: solid;)").arg(reinterpret_cast<std::intptr_t>(this));
-  m_styles.write(stylesheet);
-  if(stylesheet != styleSheet()) {
-    setStyleSheet(stylesheet);
-    if(m_body) {
-      m_size_hint = none;
-      updateGeometry();
+void Box::draw_rectangular_border(QPainter& painter) const {
+  painter.fillRect(rect(), m_style.m_border_color.m_top);
+  const auto& border_width = m_style.m_border_width;
+  painter.fillRect(rect().adjusted(border_width.m_left, border_width.m_top,
+    -border_width.m_right, -border_width.m_bottom), m_style.m_background_color);
+}
+
+void Box::draw_rounded_border(QPainter& painter) const {
+  painter.setPen(Qt::NoPen);
+  painter.setBrush(m_style.m_border_color.m_top);
+  auto radius = reduce_radius().m_top_left;
+  painter.drawRoundedRect(rect(), radius, radius);
+  painter.setBrush(m_style.m_background_color);
+  auto border_width = m_style.m_border_width.m_top;
+  painter.drawRoundedRect(rect().adjusted(
+    border_width, border_width, -border_width, -border_width),
+    std::max(0, radius - border_width), std::max(0, radius - border_width));
+}
+
+void Box::update_border_shape() {
+  m_border_shape = [&] {
+    if(m_style.m_border_radius.are_equal() &&
+        m_style.m_border_color.are_equal()) {
+      if(m_style.m_border_radius.m_top_left > 0) {
+        if(m_style.m_border_width.are_equal()) {
+          return BorderShape::ROUNDED_RECTANGLE;
+        }
+      } else {
+        return BorderShape::RECTANGLE;
+      }
     }
+    return BorderShape::COMPLEX;
+  }();
+}
+
+double Box::radius_reduction_factor(const BorderRadius& radius) const {
+  return std::min(1.0,
+    std::min(static_cast<double>(width()) /
+      (radius.m_top_left + radius.m_top_right),
+    std::min(static_cast<double>(height()) /
+      (radius.m_top_right + radius.m_bottom_right),
+    std::min(static_cast<double>(width()) /
+      (radius.m_bottom_left + radius.m_bottom_right),
+      static_cast<double>(height()) /
+      (radius.m_bottom_left + radius.m_top_left)))));
+}
+
+Box::BorderRadius Box::reduce_radius() const {
+  auto radius = m_style.m_border_radius;
+  auto factor = radius_reduction_factor(radius);
+  while(factor < 1.0f) {
```

Comment:
I don't understand why the while loop is needed here. Sometimes, it may cause a infinite loop.